docs/dev/guidelines/server-tests.rst

```diff
@@ -0,0 +1,180 @@
+* Make use of `JUnit 5 <https://junit.org/junit5/docs/current/user-guide/#writing-tests>`__ features such as parameterized tests.
+* Follow `best practices <https://www.baeldung.com/spring-tests>`__ related to spring testing.
+
+For a more detailed overview check out `modern best testing practices <https://phauer.com/2019/modern-best-practices-testing-java/>`__.
+
+
+1. Assert using the most specific overload method
+==================================================
+
+When asserting in server tests, use ``assertThat`` from the `AssertJ <https://github.com/assertj/assertj>`__ library. Another assertion statement, such as ``isEqualTo()``, **must**  follow the call. Using specific assertion statements rather than always expecting boolean values is best practice.
+
+For example, instead of
+
+.. code-block:: java
+
+    assertThat(submissions.size()).isEqualTo(1);
+    assertThat(submissionInDb.isPresent()).isTrue();
+    assertThat(submissionInDb.get().getFilePath().contains("ffile.png")).isTrue();
+
+use the built-in assertions directly:
+
+.. code-block:: java
+
+    assertThat(submissions).hasSize(1);
+    assertThat(submissionInDb).isPresent();
+    assertThat(submissionInDb.get().getFilePath()).contains("ffile.png");
+
+These assertions provide better error messages when they fail and improve the code readability. However, not all methods are suitable for this type of assertion.
+If the ``isTrue`` assertion is unavoidable, specify a custom error message using the ``as`` keyword:
+
+.. code-block:: java
+
+    assertThat(submission.isSubmittedInTime()).as("submission was not in time").isTrue();
+
+For more information, please read `the AssertJ documentation <https://assertj.github.io/doc/#assertj-core-assertions-guide>`__, especially the `section about avoiding incorrect usage <https://assertj.github.io/doc/#assertj-core-incorrect-usage>`__.
+
+
+2. ArchUnit
+===========
+Use the `ArchUnit <https://www.archunit.org/>`__ library to prevent the unintentional inclusion of unnecessary packages. We use the library to enforce consistency in the code base.
+Here is a simple ArchUnit test using an ArchRule to forbid JUnit assertions (in favor of AssertJ ones).
+
+.. code-block:: java
+
+    @Test
+    void testNoJunitJupiterAssertions() {
+        ArchRule noJunitJupiterAssertions = noClasses().should().dependOnClassesThat().haveNameMatching("org.junit.jupiter.api.Assertions");
+
+        noJunitJupiterAssertions.check(testClasses);
+    }
+
+We first define the ArchRule prohibiting the JUnit assertion package and then enforce it in test classes.
+
+
+3. Counting database query calls within tests
+==============================================
+
+It's possible to write tests checking how many database accesses a REST call performs. These tests ensure that code changes don't lead to more database calls or remind developers if they do, which is especially important for commonly used endpoints.
```

Comment:
While often used for complete REST calls, this is not limited to rest class. We can scope this to every code block wanted by the developer.

Not sure if `unintentional performance decrements` is the best formulation here, but we should shortly explain why we want to count database requests.
```suggestion
It's possible to write tests checking how many database accesses an operation performs. These tests ensure that code changes don't lead to unintentional performance decrements and remind developers if they do, which is especially important for commonly used functionality.
```